test/RoboCop.ts

```diff
@@ -132,32 +133,37 @@ describe("RoboCop", () => {
       await network.provider.send("evm_setAutomine", [false]);
       const tx1 = await roboCop.connect(ruleMakerWallet).createRule([passingTrigger], [executableAction]);
       const tx2 = await roboCop.connect(ruleMakerWallet).createRule([passingTrigger], [executableAction]);
-      // different user, so this 3rd rule should work
-      const tx3 = await roboCop.connect(ruleMakerWallet).createRule([passingTrigger], [executableAction]);
+
       await network.provider.send("evm_mine", []);
       await network.provider.send("evm_setAutomine", [true]);
+      // different block time, so this 3rd rule should work
+      await expect(roboCop.connect(ruleMakerWallet).createRule([passingTrigger], [executableAction])).to.emit(
+        roboCop,
+        "Created"
+      );
 
       try {
         await tx1.wait();
       } catch {
         expect.fail();
       }
 
-      try {
-        await tx2.wait();
-        // cant figure how to confirm whether the error is a duplicate rule error.
-        // this will suffice for now.
-        expect.fail();
-      } catch (err) {
-        /* pass */
-      }
-
-      try {
-        await tx3.wait();
-      } catch (err) {
-        // this acts as a control to ensure the error wasnt due to the evm_mine stuff
-        expect.fail();
-      }
+      const trace = await network.provider.send("debug_traceTransaction", [
+        tx2.hash,
+        {
+          disableMemory: true,
+          disableStack: true,
+          disableStorage: true,
+        },
+      ]);
+
+      expect(trace.failed).to.be.equal(true);
+
+      // checks if error is duplicated rule, its in raw form as I don't know how to abi decode
+      // 0x4475706c69636174652052756c65 is Duplicate Rule in hex
+      expect(trace.returnValue).to.be.equal(
```

Comment:
you can use something like    `ethers.utils.defaultAbiCoder.encode(["address", "uint8", "uint256"], [ETH_ADDRESS, GT, 0])`  to do an abi encode and then compare the values. But you need to know the structure of the data returned.

Reply:
I did try `ethers.utils.defaultAbiCoder.decode(["createRule((address,uint8,bytes)[],(address,bytes,(uint8,address)[],(uint8,address)[])[])"], trace.returnValue)` and `ethers.utils.defaultAbiCoder.decode(["tuple(address, uint8, bytes)[]", "tuple(address, bytes, tuple(uint8, address)[], tuple(uint8, address)[])[]"], trace.returnValue)` but it didn't work. Probably my understanding of abi encoding is still not there. Will continue to investigate in it